### ISSUE TITLE
JVM crash when using ASGCT (async-profiler)

### ISSUE THREAD

**Issue by jbachorik** [eclipse-openj9/openj9#15794]
Java -version output
--------------------

openjdk version "18.0.2" 2022-07-19
IBM Semeru Runtime Open Edition 18.0.2.0 (build 18.0.2+9)
Eclipse OpenJ9 VM 18.0.2.0 (build openj9-0.33.0, JRE 18 Linux amd64-64-Bit Compressed References 20220804_30 (JIT enabled, AOT enabled)
OpenJ9   - 04a55b45b
OMR      - b58aa2708
JCL      - dff9c97e1c7 based on jdk-18.0.2+9)

Summary of problem
------------------

Running with async-profiler (which in turn is using the semi-public AsyncGetCallTrace function to obtain stacktraces) will crash the JVM occasionally.

The steps to repeat:
1. Download Reanaissance Benchmark suite - https://github.com/renaissance-benchmarks/renaissance/releases/download/v0.14.1/renaissance-gpl-0.14.1.jar
2. Download async-profiler 2.8.3 - https://github.com/jvm-profiling-tools/async-profiler/releases/tag/v2.8.3
3. Extract async-profiler archive to <some_path>/async-profiler
4. Run following script
```sh
for i in {0..20}; do
    echo "== Iteration ${i}"
    java  -agentpath:<some_path>/async-profiler/build/libasyncProfiler.so=start,flat=100,interval=10ms,traces=1,event=cpu -jar <path_to>/renaissance-gpl-0.14.1.jar finagle-http --run-seconds 120
  done
```

The above mentioned steps will bring the JVM to crash sooner or later.

**Comment by jbachorik** [comment]
[Zipped log files](https://github.com/eclipse-openj9/openj9/files/9451365/crashes.zip)

**Comment by pshipton** [comment]
@tajila @gacholio fyi. Tentatively added to 0.35, but it can move out.

**Comment by gacholio** [comment]
```
4XENATIVESTACK               pool_newElement.cold+0x0 (0x00007F15707A1EF5 [libj9prt29.so+0xcef5])
4XENATIVESTACK                (0x00007F1570E7C520 [libc.so.6+0x42520])
4XENATIVESTACK               getJNIMethodID+0x14 (0x00007F1570866E84 [libj9vm29.so+0x43e84])
```
This is not supposed to occur - method IDs are pre-filled precisely to avoid this situation. This occurs when a class is loaded, provided the JVMTI ClassLoad event is enabled (which is a requirement for using ASGCT). An assertion to check this was added in https://github.com/eclipse-openj9/openj9/pull/14308

**Comment by gacholio** [comment]
This looks similar to the issues reported in https://github.com/eclipse-openj9/openj9/issues/13838

For whatever reason, it appears the stack-allocated buffer used by async-profiler causes problems with OpenJ9 (perhaps something in the way our signal handler is implemented).

**Comment by gacholio** [comment]
#13838 proposed a modification to async-profiler to work around the issue, but it was never committed to the project.

**Comment by jbachorik** [comment]
@gacholio I think this particular failure can not be resolved in async-profiler.

This is the interesting part
```
4XENATIVESTACK               pool_newElement.cold+0x0 (0x00007F15707A1EF5 [libj9prt29.so+0xcef5])
4XENATIVESTACK                (0x00007F1570E7C520 [libc.so.6+0x42520])
4XENATIVESTACK               getJNIMethodID+0x14 (0x00007F1570866E84 [libj9vm29.so+0x43e84])
4XENATIVESTACK               jvmtiInternalGetStackTraceIteratorExtended+0x45 (0x00007F15701020C5 [libj9jvmti29.so+0x110c5])
4XENATIVESTACK               walkFrame+0x1d8 (0x00007F15708A00F8 [libj9vm29.so+0x7d0f8])
4XENATIVESTACK               walkStackFrames+0x4ac (0x00007F15708A0CAC [libj9vm29.so+0x7dcac])
4XENATIVESTACK               jvmtiInternalGetStackTraceExtended+0x103 (0x00007F15700FF023 [libj9jvmti29.so+0xe023])
4XENATIVESTACK               jvmtiGetStackTraceExtended+0x1c7 (0x00007F1570101027 [libj9jvmti29.so+0x10027])
```

As you can see there is allocation happening in `getJNIMethodID()` which is in turn called from the ASGCT implementation (through several intermediaries).

The code in `getJNIMethodID()` looks like this
```c++
	/* If the table exists already, check it without mutex */

	jniIDs = declaringClass->jniIDs;
	if (jniIDs != NULL) {
		id = (J9JNIMethodID*)(jniIDs[methodIndex]);
		if (id != NULL) {
			return id;
		}
	}

#ifdef J9VM_THR_PREEMPTIVE
	omrthread_monitor_enter(vm->jniFrameMutex);
#endif

	jniIDs = ensureJNIIDTable(currentThread, declaringClass);
	if (jniIDs == NULL) {
		id = NULL;
	} else {
		id = (J9JNIMethodID*)(jniIDs[methodIndex]);
		if (id == NULL) {
			id = (J9JNIMethodID*)pool_newElement(declaringClass->classLoader->jniIDs);
			if (id != NULL) {
				initializeMethodID(currentThread, id, method);
				issueWriteBarrier();
				jniIDs[methodIndex] = id;
			}
		}
	}

#ifdef J9VM_THR_PREEMPTIVE
	omrthread_monitor_exit(vm->jniFrameMutex);
#endif
```

This looks like the `jniIDs` table is not initialized for some classes. I have added some trivial debug ouptut and rerunning my reproducer to figure out on which class this would choke.

**Comment by gacholio** [comment]
The code above is not from the signal handler - it most likely indicates memory corruption. In the mentioned issue, it was found that removing the call to ASGCT entirely would still result in a crash.

The apparent lack of prefilling the table may be another symptom of the corruption.

**Comment by jbachorik** [comment]
Yes, you are right. In fact, async-profiler is not invoking ASGCT from a signal handler, it turns out - as opposed to the hotspot version I am more familiar with.

Signal handler is used only to ask for the thread to be 'halted for inspection' (https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/j9StackTraces.cpp#L171) but the actual stack sampling is then done in a dedicated external thread. The signal handler itself is quite simple - https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/perfEvents_linux.cpp#L697

I did a quick hack running `loadAllMethodIds()` unconditionally (https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/vmEntry.cpp#L224) upon the startup and this seems to be preventing this crash. Could be just remediating the syndrome, though.

**Comment by jbachorik** [comment]
> https://github.com/eclipse-openj9/openj9/issues/13838 proposed a modification to async-profiler to work around the issue, but it was never committed to the project.

FWIW, async-profiler is not calling ASGCT from a signal handler any more (as per my previous comment).

**Comment by jbachorik** [comment]
> I did a quick hack running loadAllMethodIds() unconditionally (https://github.com/jvm-profiling-tools/async-profiler/blob/master/src/vmEntry.cpp#L224) upon the startup and this seems to be preventing this crash. Could be just remediating the syndrome, though.

FYI, the crash just became more rare - the workaround is definitely not fixing it :(

**Comment by gacholio** [comment]
The OpenJ9-specific modifications appear to have been done with no consultation with this team - I would not expect them to work as they are using OpenJ9 internal flags in ways that are not sanctioned.

@tajila @pshipton

**Comment by pshipton** [comment]
@gacholio are you going to provide more details about the problem(s) you spotted?

**Comment by gacholio** [comment]
The code is setting and clearing `J9_HALT_THREAD_INSPECTION` in the target thread without the necessary handshaking.

This may appear to work on threads which are already blocked (as most java threads usually are).

**Comment by gacholio** [comment]
Also, using structures and constants from our headers means that if the field offsets or constant values change, the agent would have to be recompiled.

**Comment by pshipton** [comment]
Should we create a jvmti extension that can be used?

**Comment by gacholio** [comment]
VM access is complex enough already I think.

The previously-proposed solution appeared to work (not stack allocating buffers in the signal handler).
